### PR TITLE
[prep CDF-24745, CDF-24554, CDF-24555] 👷Prepare dump refactoring, TableWriter class.

### PR DESCRIPTION
**Pull request by doctrino**
# Description

**Context**: Toolkit supports dumping assets/timeseries to csv/parquet (and yaml). That command will be refactored to support non-asset centric resources such as Canvas as part of migration support. This PR introduces a TableWriter class with subclasses for writing to parquet/csv/yaml which will be used in the refactoring. It is pulled out in this PR to keep PR size small.

## Changelog

- [ ] Patch
- [ ] Minor
- [x] Skip

### CODE REVIEW

File: cognite_toolkit/_cdf_tk/utils/table_writers.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+        for prop in self.schema.columns:
+            pa_type = self._as_pa_type(prop.type, prop.is_array)
+            fields.append(pa.field(prop.name, pa_type, nullable=True))
+        return pa.schema(fields)
+
+    @staticmethod
+    def _check_pyarrow_dependency() -> None:
+        if importlib.util.find_spec("pyarrow") is None:
+            raise ToolkitMissingDependencyError(
+                "Writing to parquet requires pyarrow. Install with 'pip install cognite-toolkit[table]'"
```
Comment (ronpal) [inline cognite_toolkit/_cdf_tk/utils/table_writers.py:165]: This fails in some terminals, enclosing in single or double quotes may be necessary: 

```suggestion
                "Writing to parquet requires pyarrow. Install with 'pip install \"cognite-toolkit[table]\"'"
```

**Review by ronpal — APPROVED** [review]

**Review by ronpal — APPROVED** [review]

File: cognite_toolkit/_cdf_tk/utils/table_writers.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+from cognite_toolkit._cdf_tk.utils import humanize_collection, to_directory_compatible
+from cognite_toolkit._cdf_tk.utils.file import yaml_safe_dump
+
+if TYPE_CHECKING:
+    import pyarrow as pa
+    import pyarrow.parquet as pq
+
+FileFormat: TypeAlias = Literal["csv", "parquet", "yaml"]
+DataType: TypeAlias = Literal["string", "integer", "float", "boolean", "datetime", "date", "time", "json"]
+Rows: TypeAlias = list[dict[str, Any]]
```
Comment (larshagencognite) [inline cognite_toolkit/_cdf_tk/utils/table_writers.py:22]: is `Any` correct here, or should we use a nested json type?

File: tests/test_unit/test_cdf_tk/test_utils/test_table_writers.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        with CSVWriter(example_schema, output_dir) as writer:
+            writer.write_rows([("group1", [{"column1": "value1", "column2": 1, "column3": 1.0, "column4": ["value"]}])])
+            writer.write_rows([("group1", [{"column1": "value2", "column3": 2.0}])])
+
+        csv_file = list(output_dir.rglob("*.csv"))
+        assert len(csv_file) == 1
+        actual_file = csv_file[0].read_text()
+        assert actual_file == ("column1,column2,column3,column4\nvalue1,1,1.0,['value']\nvalue2,,2.0,\n")
+
+    def test_write_yaml(self, example_schema: Schema, tmp_path: Path) -> None:
```
Comment (larshagencognite) [inline tests/test_unit/test_cdf_tk/test_utils/test_table_writers.py:44]: Should verify that we handle corner cases correctly for yaml, like `false` vs `"false"`. Do we support yaml 1.1 parsers or only >1.2? Older parsers might treat `yes/no/on/off/y/n` etc. as a boolean, so it needs quotes in more cases.

File: cognite_toolkit/_cdf_tk/utils/table_writers.py (hunk 2 above)
Comment (doctrino) [reply cognite_toolkit/_cdf_tk/utils/table_writers.py:22]: Did a more specific type, however, has to keep the `object` and `list` as we do support `json`.

**Review by ronpal — DISMISSED** [review]

File: cognite_toolkit/_cdf_tk/utils/table_writers.py [hunk 4]
```diff
@@ -18,8 +19,8 @@
     import pyarrow.parquet as pq
 
 FileFormat: TypeAlias = Literal["csv", "parquet", "yaml"]
-DataType: TypeAlias = Literal["string", "integer", "float", "boolean", "datetime", "date", "time", "json"]
-Rows: TypeAlias = list[dict[str, Any]]
+DataType: TypeAlias = Literal["string", "integer", "float", "boolean", "json"]
+Rows: TypeAlias = list[dict[str, str | int | float | bool | datetime | date | object | list | None]]
```
Comment (larshagencognite) [inline cognite_toolkit/_cdf_tk/utils/table_writers.py:23]: Is a recursive definition of a json value more correct here?

Comment (larshagencognite) [reply cognite_toolkit/_cdf_tk/utils/table_writers.py:23]: where `JsonVal: TypeAlias = None | str | int | float | bool | dict[str, "JsonVal"] | list["JsonVal"]`

File: cognite_toolkit/_cdf_tk/utils/table_writers.py [hunk 5]
```diff
@@ -20,7 +19,8 @@
 
 FileFormat: TypeAlias = Literal["csv", "parquet", "yaml"]
 DataType: TypeAlias = Literal["string", "integer", "float", "boolean", "json"]
-Rows: TypeAlias = list[dict[str, str | int | float | bool | datetime | date | object | list | None]]
+JsonVal: TypeAlias = None | str | int | float | bool | dict[str, "JsonVal"] | list["JsonVal"]
+Rows: TypeAlias = list[dict[str, str | int | float | bool | JsonVal | None]]
```
Comment (larshagencognite) [inline cognite_toolkit/_cdf_tk/utils/table_writers.py:23]: I think the fact that this expands to duplicates of identical types is a bit strange, but not blocking.
